src/modules/lua/script_lua.c

```diff
@@ -189,451 +221,388 @@ void *luaGetFromRegistry(lua_State *lua, const char *name) {
+    while (sp <= end && strchr(cset, *sp)) sp++;
+    while (ep > sp && strchr(cset, *ep)) ep--;
+    len = (ep - sp) + 1;
+    if (s != sp) memmove(s, sp, len);
+    s[len] = '\0';
+    return s;
 }
 
+/* This function is used in order to push an error on the Lua stack in the
+ * format used by server.pcall to return errors, which is a lua table
+ * with an "err" field set to the error string including the error code.
+ * Note that this table is never a valid reply by proper commands,
+ * since the returned tables are otherwise always indexed by integers, never by strings.
+ *
+ * The function takes ownership on the given err_buffer. */
+static void luaPushErrorBuff(lua_State *lua, const char *err_buffer) {
+    char *msg;
+    char *error_code;
 
-static void
-redisProtocolToLuaType_BulkString(void *ctx, const char *str, size_t len, const char *proto, size_t proto_len) {
-    UNUSED(proto);
-    UNUSED(proto_len);
-    if (!ctx) {
-        return;
-    }
-
-    lua_State *lua = ctx;
-    if (!lua_checkstack(lua, 1)) {
-        /* Increase the Lua stack if needed, to make sure there is enough room
-         * to push elements to the stack. On failure, exit with panic. */
-        serverPanic("lua stack limit reach when parsing server.call reply");
+    /* If debugging is active and in step mode, log errors resulting from
+     * server commands. */
+    if (ldbIsEnabled()) {
+        char *msg = lm_asprintf("<error> %s", err_buffer);
+        ldbLogCString(msg);
+        ValkeyModule_Free(msg);
     }
-    lua_pushlstring(lua, str, len);
-}
 
-static void redisProtocolToLuaType_Status(void *ctx, const char *str, size_t len, const char *proto, size_t proto_len) {
-    UNUSED(proto);
-    UNUSED(proto_len);
-    if (!ctx) {
-        return;
+    char *final_msg = NULL;
+    /* There are two possible formats for the received `error` string:
+     * 1) "-CODE msg": in this case we remove the leading '-' since we don't store it as part of the lua error format.
+     * 2) "msg": in this case we prepend a generic 'ERR' code since all error statuses need some error code.
+     * We support format (1) so this function can reuse the error messages used in other places.
+     * We support format (2) so it'll be easy to pass descriptive errors to this function without worrying about format.
+     */
+    if (err_buffer[0] == '-') {
+        /* derive error code from the message */
+        char *err_msg = strstr(err_buffer, " ");
+        if (!err_msg) {
+            msg = lm_strcpy(err_buffer + 1);
+            error_code = lm_strcpy("ERR");
+            final_msg = lm_asprintf("%s %s", error_code, msg);
+            ValkeyModule_Free(error_code);
+        } else {
+            *err_msg = '\0';
+            msg = lm_strcpy(err_msg + 1);
+            msg = lm_strtrim(msg, "\r\n");
+            error_code = lm_strcpy(err_buffer + 1);
+            final_msg = lm_asprintf("%s %s", error_code, msg);
+            ValkeyModule_Free(error_code);
```

Comment:
We can avoid a temp allocation.

```suggestion
            final_msg = lm_asprintf("%s %s", err_buffer + 1, msg);
```